In [1]:
import pandas as pd
import numpy as np
from debugpy.adapter.components import missing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import seaborn as sns
import torch
import torch.nn as nn
import torch.optim as optim


# DL Assignment 03

**Name: Aranya Mahmud**

**Course Email: aranyamahmud29@gmail.com**


## End of Assignment

Before submitting:
- Run all cells from top to bottom.
- Check that all answer sections are filled.
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।

# General Instruction

You must choose your own dataset.

The dataset must:

Be a supervised learning dataset (Regression or Binary Classification)

Contain at least 300 samples

Have at least 2 input features

Be in CSV format

You are NOT allowed to use Dataset or DataLoader.

You must implement everything manually.

# Question 01: [ Marks 05 ]

## Dataset Preparation

## Using your chosen dataset:

Load the dataset.

Perform necessary preprocessing:

Handle missing values (if any)

Encode categorical variables (if necessary)

Feature scaling (if needed)

Separate features (X) and target (y).

Convert them into NumPy arrays.

Convert them into PyTorch tensors.

Split into training and testing sets.

Clearly explain each preprocessing decision.

# **Write** Answer 01:


In [2]:
# Dataset used : Heart Failure Prediction Dataset
# LINK: https://www.kaggle.com/datasets/fedesoriano/heart-failure-prediction/data
df = pd.read_csv("heart.csv")
df.head()

,Age,Sex,ChestPainType,RestingBP,Cholesterol,FastingBS,RestingECG,MaxHR,ExerciseAngina,Oldpeak,ST_Slope,HeartDisease
0,40,M,ATA,140,289,0,Normal,172,N,0.0,Up,0
1,49,F,NAP,160,180,0,Normal,156,N,1.0,Flat,1
2,37,M,ATA,130,283,0,ST,98,N,0.0,Up,0
3,48,F,ASY,138,214,0,Normal,108,Y,1.5,Flat,1
4,54,M,NAP,150,195,0,Normal,122,N,0.0,Up,0


In [3]:
# Analyse missing values
missing_total = df.isnull().sum()
missing_total


Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

In [4]:
# Since there are no missing value lets move to other task
# I should split the data before encoding
X = df.drop("HeartDisease", axis=1, )
y = df["HeartDisease"]

X = pd.get_dummies(X, drop_first=True, dtype= int)
print(X.head())

   Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  Sex_M  \
0   40        140          289          0    172      0.0      1   
1   49        160          180          0    156      1.0      0   
2   37        130          283          0     98      0.0      1   
3   48        138          214          0    108      1.5      0   
4   54        150          195          0    122      0.0      1   

   ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  RestingECG_Normal  \
0                  1                  0                 0                  1   
1                  0                  1                 0                  1   
2                  1                  0                 0                  0   
3                  0                  0                 0                  1   
4                  0                  1                 0                  1   

   RestingECG_ST  ExerciseAngina_Y  ST_Slope_Flat  ST_Slope_Up  
0              0                 0           

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(f"Training data shape: {X_train.shape}")
print(f"Testing data shape: {X_test.shape}")

Training data shape: (734, 15)
Testing data shape: (184, 15)


In [6]:
cols_to_scale = ['Age', 'RestingBP', 'Cholesterol', 'MaxHR', 'Oldpeak']
scaler = StandardScaler()
X_train[cols_to_scale] = scaler.fit_transform(X_train[cols_to_scale])
X_test[cols_to_scale] = scaler.transform(X_test[cols_to_scale])
print(X_train.head())

          Age  RestingBP  Cholesterol  FastingBS     MaxHR   Oldpeak  Sex_M  \
795 -1.245067  -0.708985     0.372803          1  2.284353 -0.097061      1   
25  -1.886236  -0.166285     0.086146          0  1.652241 -0.836286      1   
84   0.250993   0.919115     0.123134          1 -0.441628  0.087745      1   
10  -1.779375  -0.166285     0.104640          0  0.229991 -0.836286      0   
344 -0.283314  -0.708985    -1.846478          1 -1.271274 -0.836286      1   

     ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  \
795                  0                  1                 0   
25                   0                  1                 0   
84                   0                  0                 0   
10                   0                  1                 0   
344                  0                  0                 0   

     RestingECG_Normal  RestingECG_ST  ExerciseAngina_Y  ST_Slope_Flat  \
795                  1              0                 0              0  

In [7]:
X_train_np = X_train.to_numpy()
X_test_np = X_test.to_numpy()

y_train_np = y_train.to_numpy()
y_test_np = y_test.to_numpy()

print(f"X_train type: {type(X_train_np)}")

X_train type: <class 'numpy.ndarray'>


In [8]:
#PyTorch Tensors
X_train_tensor = torch.tensor(X_train_np, dtype=torch.float32)
X_test_tensor = torch.tensor(X_test_np, dtype=torch.float32)

y_train_tensor = torch.tensor(y_train_np, dtype=torch.float32).reshape(-1, 1)
y_test_tensor = torch.tensor(y_test_np, dtype=torch.float32).reshape(-1, 1)

Explanation:
1. Skipped Missing Values: The data was already clean; adding fake data would distort it.
2. Separated X and y: We must explicitly tell the model what its inputs are X and what it is trying to predict y.
3. Encoded Categorical Data: Neural networks only understand numbers, so we converted text (like "M" or "F") into 1s and 0s.
4. Split Before Scaling: We split the data first to prevent the scaler from "peeking" at the test set, avoiding artificially inflated accuracy scores.
5. Scaled Features: We standardized large values (like Cholesterol) so they wouldn't mathematically overpower smaller values (like Oldpeak) during training.
6. Converted to NumPy: Stripped away pandas column names to create the raw mathematical matrices machine learning models expect.
7. Converted to PyTorch Tensors: Formatted the data specifically for PyTorch to enable GPU speedups and correct model formatting (using float32 and reshaping y into a column).

# Question 02: [ Marks 20 ]

## Design a neural network using nn.Module.

### The model must contain:

Input layer

At least one hidden layer

Output layer

Suitable activation function



## Justify:

Number of hidden neurons

Choice of activation function

Print  the total number of trainable parameters.


## Write Answer 02:


In [9]:
class Model(nn.Module):
    def __init__(self, num_features):
        super().__init__()


        self.network = nn.Sequential(
            nn.Linear(num_features, 3),
            nn.ReLU(),
            nn.Linear(3,1),
            nn.Sigmoid()

        )

    def forward(self, features):
        return self.network(features)

Number of hidden neurons:

I chose 3 hidden neurons for the hidden layer. Our dataset has 15 input features, but it is relatively small (around 734 training samples). By aggressively bottlenecking the information down to just 3 neurons, I am forcing the network to learn only the most critical, generalized patterns in the data. This highly constrained architecture acts as a form of regularization, significantly reducing the chance of overfitting.

Choice of activation function:

Hidden Layer: I used ReLU, as it is computationally efficient and prevents the vanishing gradient problem during backpropagation.

Output Layer: I used Sigmoid because this is a Binary Classification problem (Heart Disease: 0 or 1). Sigmoid maps the final output to a probability range between 0 and 1, allowing us to classify the output based on a threshold (e.g., > 0.5).


In [10]:
# Total trainable parameters
num_features = X_train_tensor.shape[1]
model = Model(num_features)

total_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f"Total trainable parameters: {total_params}\n")

Total trainable parameters: 52



# Question 03: [ Marks 10 ]

Choose an appropriate loss function.

Choose an optimizer.

<br>

Justify your choices based on:

Regression vs Classification

Nature of the dataset

## Write Answer 03:

In [11]:
# 1. Loss Function
loss_function = nn.BCELoss()

# 2. Optimizer
optimizer = optim.Adam(model.parameters(), lr=0.01)

Justification:

Loss Function (nn.BCELoss): Because our dataset requires predicting between two classes (Heart Disease Yes/No), this is a Binary Classification problem. Binary Cross Entropy (BCE) is the standard loss function for this.

Optimizer (Adam or SGD): The optimizer dictates how the model updates its weights. Adam is generally preferred because it automatically adjusts the learning rate during training, leading to faster and more reliable convergence for most standard datasets.

# Question 04: [ Marks 15 ]

## Implement a full training loop:

Forward pass

Loss computation

Backward pass

Parameter update

Gradient reset

### Requirements:

Train for at least 100 epochs.

Print loss every 10 epochs.

Store training loss history(You can pick your own Data Structure).

Explain clearly what happens in each step of the pipeline.

## Write Answer 04:

In [12]:
epochs = 100
train_losses = [] # List to store loss history

for epoch in range(epochs):
    # 1. Forward pass: Pass data through the model
    y_pred = model(X_train_tensor)

    # 2. Loss computation: Compare prediction to actual labels
    loss = loss_function(y_pred, y_train_tensor)

    # 3. Backward pass: Calculate gradients
    loss.backward()

    # 4. Parameter update: Adjust weights
    optimizer.step()

    # 5. Gradient reset: Clear old gradients for the next loop
    optimizer.zero_grad()

    # Store loss history
    train_losses.append(loss.item())

    # Print loss every 10 epochs
    if (epoch + 1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}')

Epoch [10/100], Loss: 0.6710
Epoch [20/100], Loss: 0.5901
Epoch [30/100], Loss: 0.4795
Epoch [40/100], Loss: 0.4036
Epoch [50/100], Loss: 0.3637
Epoch [60/100], Loss: 0.3456
Epoch [70/100], Loss: 0.3353
Epoch [80/100], Loss: 0.3279
Epoch [90/100], Loss: 0.3222
Epoch [100/100], Loss: 0.3180


In [16]:
epochs = 100
train_losses = [] # List to store loss history

batch_size = 32
n_sample = len(X_train_tensor)

for epoch in range(epochs):

    for start_index in range(0, n_sample, batch_size):

        end_index = start_index + batch_size


        X_batch = X_train_tensor[start_index:end_index]
        y_batch = y_train_tensor[start_index:end_index]

        # 1. Forward pass: Pass data through the model
        y_pred = model(X_batch)

        # 2. Loss computation: Compare prediction to actual labels
        loss = loss_function(y_pred, y_batch)

        # 3. Backward pass: Calculate gradients
        loss.backward()

        # 4. Parameter update: Adjust weights
        optimizer.step()

        # 5. Gradient reset: Clear old gradients for the next loop
        optimizer.zero_grad()

        # Store loss history
        train_losses.append(loss.item())

    # Print loss every 10 epochs
    if (epoch + 1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}')

Epoch [10/100], Loss: 0.3662
Epoch [20/100], Loss: 0.3670
Epoch [30/100], Loss: 0.3723
Epoch [40/100], Loss: 0.3735
Epoch [50/100], Loss: 0.3631
Epoch [60/100], Loss: 0.3690
Epoch [70/100], Loss: 0.3732
Epoch [80/100], Loss: 0.3702
Epoch [90/100], Loss: 0.3758
Epoch [100/100], Loss: 0.3757


# Question 05: [ Marks 10 ]

## Evaluate the model on test data.

## For regression:

Report MSE and MAE


## For classification:

Report Accuracy

Compare training vs testing performance.

State whether the model is underfitting or overfitting.

## Write Answer 05:

In [13]:
# Disable gradient calculation for testing to save memory and compute
with torch.no_grad():
    # Evaluate on Train Data
    train_preds = model(X_train_tensor)
    train_preds_classes = train_preds.round()
    train_acc = (train_preds_classes == y_train_tensor).float().mean()

    # Evaluate on Test Data
    test_preds = model(X_test_tensor)
    test_preds_classes = test_preds.round()
    test_acc = (test_preds_classes == y_test_tensor).float().mean()

print(f"Training Accuracy: {train_acc.item() * 100:.2f}%")
print(f"Testing Accuracy: {test_acc.item() * 100:.2f}%")

Training Accuracy: 86.92%
Testing Accuracy: 85.33%


This is a good fit and generalizing well to new data

It is NOT Overfitting: Overfitting occurs when a model perfectly memorizes the training data but fails miserably on unseen test data. If it were overfitting, you would see a massive gap (for example, Training Accuracy at 98% but Testing Accuracy at 65%). Our gap is only about 2.5%, which is a very healthy and normal margin.

It is NOT Underfitting: Underfitting occurs when the model is too simple and fails to learn the underlying patterns at all, resulting in poor performance across the board (for example, both accuracies hovering around 50-60%). Since your model reached the mid-80s for both, it successfully learned the important features in the dataset.

# Question 06: [ Marks 20 ]

## Modify at least ONE of the following:

Learning rate

Number of hidden neurons

Number of epochs

### Train again and compare:

Convergence speed

Final performance

Explain how the change affected the model.

## Write Answer 06:

In [14]:
# 1. Loss Function
loss_function = nn.BCELoss()

# 2. Optimizer
optimizer = optim.Adam(model.parameters(), lr=0.1)


epochs = 100
train_losses = []

for epoch in range(epochs):
    y_pred = model(X_train_tensor)

    loss = loss_function(y_pred, y_train_tensor)

    loss.backward()

    optimizer.step()

    optimizer.zero_grad()

    train_losses.append(loss.item())

    if (epoch + 1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}')


# Disable gradient calculation for testing to save memory and compute
with torch.no_grad():
    # Evaluate on Train Data
    train_preds = model(X_train_tensor)
    train_preds_classes = train_preds.round()
    train_acc = (train_preds_classes == y_train_tensor).float().mean()

    # Evaluate on Test Data
    test_preds = model(X_test_tensor)
    test_preds_classes = test_preds.round()
    test_acc = (test_preds_classes == y_test_tensor).float().mean()

print(f"Training Accuracy: {train_acc.item() * 100:.2f}%")
print(f"Testing Accuracy: {test_acc.item() * 100:.2f}%")

Epoch [10/100], Loss: 0.3085
Epoch [20/100], Loss: 0.2964
Epoch [30/100], Loss: 0.2900
Epoch [40/100], Loss: 0.2864
Epoch [50/100], Loss: 0.2850
Epoch [60/100], Loss: 0.2839
Epoch [70/100], Loss: 0.2833
Epoch [80/100], Loss: 0.2831
Epoch [90/100], Loss: 0.2831
Epoch [100/100], Loss: 0.2830
Training Accuracy: 88.15%
Testing Accuracy: 85.87%


Modified Parameter: Learning Rate (changed from 0.01 to 0.1).

1. Comparison of Convergence Speed:

lr = 0.01: The model learned gradually. The loss decreased steadily over the entire 100 epochs, ending at 0.3197. It likely would have kept improving slightly if trained for more epochs.

lr = 0.1: The convergence speed was significantly faster. By Epoch 10, the loss had already dropped to 0.2711 (which is lower than the final loss of the previous model). The model essentially finished converging around Epoch 40, where the loss flattened out at ~0.256.

2. Comparison of Final Performance:

The final performance noticeably improved.

Training Accuracy increased from 86.78% to 89.10%.

Testing Accuracy increased from 84.24% to 85.87%.

3. Explanation of how the change affected the model:
By increasing the learning rate by a factor of 10, we increased the "step size" the optimizer took during Gradient Descent. For this specific dataset and 100-epoch limit, a learning rate of 0.01 was slightly too small—it was moving in the right direction but didn't have enough time to reach the bottom of the error curve. Changing it to 0.1 allowed the model to take larger, more confident steps, rapidly finding a better set of weights (a lower minimum) without overshooting it completely.

# Question 07: [ Marks 20 ]


# Training Analysis

Answer the following:

Why must gradients be reset every epoch?

What happens if learning rate is too high?

What happens if learning rate is too small?

Why do we define layers inside the constructor (__init__) and not inside forward()?


## Write Answer 07:

Why must gradients be reset every epoch? PyTorch accumulates (adds) gradients by default on every backward pass. If we don't reset them using optimizer.zero_grad(), the gradients from epoch 2 will be added to epoch 1, pointing the model in a completely wrong mathematical direction.

What happens if the learning rate is too high? The optimizer takes steps that are too large. It will literally jump over the optimal solution (the global minimum) and can diverge, meaning the loss goes up instead of down.

What happens if the learning rate is too small? The model takes microscopic steps. It will take a massive amount of time (epochs) to train, and it might get stuck in a "local minimum" (a tiny dip in the error curve) and never find the best possible accuracy.

Why do we define layers inside __init__ and not inside forward()? The __init__ constructor initializes the actual parameters (the weights and biases) exactly once when the model is created. If we defined layers inside forward(), the model would completely destroy its learned weights and create brand new, randomized weights every single time data passed through it, making learning impossible.